### PR TITLE
Add tests for the Zibi extension

### PR DESCRIPTION
**Pull request by wangpc-pp**
This PR adds architectural tests for the Zibi extension.

The spec is https://github.com/riscv/zibi/blob/main/src/unpriv/zibi.adoc.

This is based on the original PR #1060.

Co-authored-by: Jiacheng Tong <tjc.challenger1024@gmail.com>

### CODE REVIEW

**Comment by wangpc-pp** [comment]
There are some failures like:
```
✗ FAILED: UDB config validation (spike-rv64-max)
  Output: work/spike-rv64-max/.validated
  Error output:
    UDB configuration validation failed for spike-rv64-max.yaml
    ℹ info    Merged architecture files written to /home/runner/.local/share/udb/0.1.15/gen/spec/_
    ℹ info    Resolved architecture files written to /home/runner/.local/share/udb/0.1.15/gen/resolved_spec/_
    ℹ info    Constructing ConfiguredArchiture for spike-rv64-max
    Config spike-rv64-max is invalid
    
      * Zibi is not a known extension
```

It seems the released UDB (v0.1.15) this PR uses doesn't have the Zibi support. @jordancarlin Can you please help with that?

**Comment by jordancarlin** [comment]
> It seems the released UDB (v0.1.15) this PR uses doesn't have the Zibi support. Jordan Carlin (@jordancarlin) Can you please help with that?

Looks like the PR adding Zibi to UDB was merged right after the last release. There are weekly UDB releases, so this should be resolved with the next release (next Monday).

File: framework/src/act/fcov/coverage/RISCV_instruction_base.svh [hunk 1]
```diff
@@ -489,6 +489,10 @@ class RISCV_instruction
     current.imm3 = get_imm(ops[offset].key);
   endfunction
 
+  virtual function void add_cimm(int offset);
+    current.cimm = get_imm(ops[offset].key);
+  endfunction
```
Comment (jordancarlin) [inline framework/src/act/fcov/coverage/RISCV_instruction_base.svh]: Let's use the existing imm2 function instead of adding a new function for cimm.

File: framework/src/act/fcov/coverage/RISCV_trace_data.svh [hunk 2]
```diff
@@ -153,6 +153,7 @@ class riscvTraceData
   bit [(XLEN-1):0]          imm;
   bit [(XLEN-1):0]          imm2;
   bit [(XLEN-1):0]          imm3;
+  bit [(XLEN-1):0]          cimm;
```
Comment (jordancarlin) [inline framework/src/act/fcov/coverage/RISCV_trace_data.svh]: You should be able to drop this after switching to imm2.

File: framework/src/act/fcov/disassemble.svh [hunk 3]
```diff
@@ -194,6 +195,8 @@ function string disassemble (logic [31:0] instrRaw);
     BLT:     $sformat(decoded, "blt %s, %s, %0d", rs1, rs2, immBType);
     BLTU:    $sformat(decoded, "bltu %s, %s, %0d", rs1, rs2, immBType);
     BNE:     $sformat(decoded, "bne %s, %s, %0d", rs1, rs2, immBType);
+    BEQI: $sformat(decoded, "beqi %s, %0d, %0d", rs1, cimm5, immBType);
+    BNEI: $sformat(decoded, "bnei %s, %0d, %0d", rs1, cimm5, immBType);
```
Comment (jordancarlin) [inline framework/src/act/fcov/disassemble.svh]: Keep this file grouped by extension.

File: generators/testgen/src/testgen/data/edges.py [hunk 4]
```diff
@@ -76,6 +76,12 @@ class IMMEDIATE_EDGES:
     # 5-bit Unsigned Immediates (Vector Instructions)
     imm_5bit_u = (0, 1, 2, 15, 16, 30, 31)
 
+    # 5-bit cimm field (BEQI, BNEI instructions in Zibi extension).
+    # In Zibi the cimm field is not sign-extended: field 0 encodes the value -1
+    # and fields 1..31 encode 1..31. The nonzero coverpoints therefore drop the
+    # field value 0 and add the effective value -1 instead (see cp_imm_edges.py).
+    imm_5bit_zibi = (0, 1, 2, 3, 4, 8, 16, 30, 31)
```
Comment (jordancarlin) [inline generators/testgen/src/testgen/data/edges.py]: Seems like it would be simpler to drop 0 from this list and add -1. Then we can use the list directly without modifying it whenever it is called.

File: generators/testgen/src/testgen/coverpoints/cp_imm_edges.py [hunk 5]
```diff
@@ -28,6 +28,11 @@ def make_cp_imm_edges(instr_name: str, instr_type: str, coverpoint: str, test_da
         edges_imm = IMMEDIATE_EDGES.imm_6bit
     elif coverpoint.endswith("_6bit_n0"):
         edges_imm = IMMEDIATE_EDGES.imm_6bit[1:]  # exclude imm=0
+    elif coverpoint.endswith("_5bit"):
+        edges_imm = IMMEDIATE_EDGES.imm_5bit
```
Comment (jordancarlin) [inline generators/testgen/src/testgen/coverpoints/cp_imm_edges.py]: Is this a remnant from an earlier attempt? Doesn't look like this is used anywhere.

File: testplans/Zibi.csv [hunk 6]
```diff
@@ -0,0 +1,3 @@
+Instruction,Type,RV32,RV64,cp_asm_count,cp_rs1,cp_rs1_edges,cr_rs1_cimm_edges,cp_offset
```
Comment (jordancarlin) [inline testplans/Zibi.csv]: Let's add `cp_imm_edges` as an additional coverpoint with `5bit_u_n0` as the variant. I know that is sort of covered already with `cr_rs1_cimm`, but we prefer to test them individually as well.

Comment (jordancarlin) [reply testplans/Zibi.csv]: Also, I think we should use the existing `cr_rs1_imm_edges` coverpoint with a variant called `5bit_u_n0_offset`. Make sure to update the coverpoint templates accordingly.

File: generators/coverage/src/covergroupgen/templates/cr_rs1_cimm_edges_offset.sv [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+        bins b_0 = {-1}; // cimm field 0 encodes the value -1
+        bins b_1 = {1};
+        bins b_2 = {2};
+        bins b_3 = {3};
+        bins b_4 = {4};
+        bins b_8 = {8};
+        bins b_16 = {16};
+        bins b_30 = {30};
+        bins b_31 = {31};
+    }
```
Comment (jordancarlin) [inline generators/coverage/src/covergroupgen/templates/cr_rs1_cimm_edges_offset.sv]: Move this into a separate coverpoint template named `cp_imm_edges_5bit_u_n0`.

**Review by jordancarlin — CHANGES_REQUESTED** [review]
Thanks for continuing this work. Here is an initial review. I will review the actual generated tests once this round of comments is resolved and UDB is updated.

File: generators/testgen/src/testgen/coverpoints/cr_rs1_imm_edges.py [hunk 8]
```diff
@@ -19,6 +19,9 @@
 @add_coverpoint_generator("cr_rs1_imm_edges")
 def make_cr_rs1_imm_edges(instr_name: str, instr_type: str, coverpoint: str, test_data: TestData) -> list[TestChunk]:
     """Generate tests for cross-product of rs1 and immediate edge values."""
+    if coverpoint.endswith("_5bit_u_n0_offset"):
+        return make_cr_rs1_imm_edges_offset(instr_name, instr_type, coverpoint, test_data)
```
Comment (jordancarlin) [inline generators/testgen/src/testgen/coverpoints/cr_rs1_imm_edges.py]: Update the function name suffix to match the coverpoint suffix.

**Review by jordancarlin — CHANGES_REQUESTED** [review]
This is looking great! I will review the generated tests and coverpoints once we update UDB and confirm this runs with the new version of UDB.

In the meantime, please add a new section for this extension to the CTP. This should include a description of the new coverpoint variants that were added and a mapping between the coverpoints and normative rules. Any of the other unpriv extensions should work as an example. This will require changes to one of the adoc files in `docs/ctp/src` and a YAML file at `coverpoints/norm/Zibi.yaml`. Reach out if you have any questions.

**Comment by jordancarlin** [comment]
ACT has been updated to use the latest version of UDB. Can you please rebase this PR? Once CI is passing, I will review again.

**Comment by wangpc-pp** [comment]
> ACT has been updated to use the latest version of UDB. Can you please rebase this PR? Once CI is passing, I will review again.

It seems we have passed (Edit: hmmm, skipped) most tests, except one:
```
ValueError: GNU assembler does not support extension(s) zibi required by 
-march=rv64i_zibi_zicsr_zifencei.
```
It is because `binutils` doesn't support `Zibi` now.

**Comment by jordancarlin** [comment]
> It is because `binutils` doesn't support `Zibi` now.

How have you been compiling the tests locally? Are you using a custom toolchain?

**Comment by wangpc-pp** [comment]
> > It is because `binutils` doesn't support `Zibi` now.
> 
> How have you been compiling the tests locally? Are you using a custom toolchain?

Yes, there is an unmerged implementation: https://github.com/sifive/riscv-binutils-gdb/tree/kitoc/zibi.

Is this blocking? Do we need to wait until binutils has supported Zibi and released? That will cost half a year.

**Comment by wangpc-pp** [comment]
> > > It is because `binutils` doesn't support `Zibi` now.
> > 
> > 
> > How have you been compiling the tests locally? Are you using a custom toolchain?
> 
> Yes, there is an unmerged implementation: https://github.com/sifive/riscv-binutils-gdb/tree/kitoc/zibi.
> 
> Is this blocking? Do we need to wait until binutils has supported Zibi and released? That will cost half a year.

@jordancarlin Hi Jordan! As far as I know, components in GNU toolchain like GCC/binutils usually don't accept unratified extensions. But it seems that we can't pass the CI as the binutils/GCC that CI is using doesn't support Zibi, so we can't merge the ACT support. And thus we can't make Zibi ratified as ACT is one of the prerequisite of ratification. Now it is an egg-chicken problem. :-(

Can we merge it if Clang tests are passed and local GCC tests are verified as well?

**Comment by jordancarlin** [comment]
@wangpc-pp The RVI freeze/ratification requirements require a PR that has been approved, not merged (for reasons like this). I will finish reviewing this in the next few days (I just need to look at the generated tests and the new additions to the CTP) and then would feel comfortable approving this for the ACT freeze milestone.

Separately, we will need to discuss how we want to handle extensions that are not yet supported in GCC. We may merge it but keep the extension disabled or may wait to merge until the assembler support is in place. We will discuss this in the next ACT SIG meeting. Either way, it will not block you from ratification. Thanks for all the great work on this PR!

**Comment by wangpc-pp** [comment]
Ping! I got a reminder that I need to get an approval quickly so that we won't exceed the timeline of ratification. Can you please help me to verify and approve this? Thanks! @jordancarlin

**Comment by wangpc-pp** [comment]
With latest commits c632915d66c02aa54cff30391920540b11fd2106 and af903f4774f8350043bcf97bd46c9ae7b63659be, the CI is passed. cc @jordancarlin

**Comment by wangpc-pp** [comment]
Ping!

File: run_tests.py
Comment (jordancarlin) [inline run_tests.py:1]: Revert the changes to this file. We discussed this in the ACT SIG yesterday, and I am going to be adding an experimental mode to ACTs soon so that we can merge unratified extensions. In the meantime, let's limit this PR to just the Zibi specific pieces.

File: generators/testgen/src/testgen/generate/unpriv.py
Comment (jordancarlin) [inline generators/testgen/src/testgen/generate/unpriv.py:1]: Revert these changes too. This will be handled with the experimental mode.

File: config/sail/sail-rv32-max-clang/run_cmd.txt
Comment (jordancarlin) [inline config/sail/sail-rv32-max-clang/run_cmd.txt:1]: The changes to these files should be reverted too.

File: docs/DeveloperGuide.md
Comment (jordancarlin) [inline docs/DeveloperGuide.md:1]: Revert

File: framework/src/act/build_plan.py
Comment (jordancarlin) [inline framework/src/act/build_plan.py:1]: Revert

File: docs/ctp/src/zi.adoc [hunk 9]
```diff
@@ -4,6 +4,28 @@
...
+[cols="3, 1, 6" options=header]
+|===
+|Modified Coverpoint|Bins|Definition
+|cp_imm_edges_5bit_u_n0|9|The Zibi comparison constants -1, 1, 2, 3, 4, 8, 16, 30, and 31. The value -1 exercises the all-zero `cimm` encoding.
+|cr_rs1_imm_edges_5bit_u_n0_offset|RV32: 216; RV64: 288|Cross-product of `rs1` edge values, `cp_imm_edges_5bit_u_n0`, and positive and negative branch offsets.
+|===
+
+The instruction-address-misaligned exception rule uses the general conditional
+branch tests in `ExceptionsSm`, which check both taken branches that trap and
+not-taken branches that do not trap.
```
Comment (jordancarlin) [inline docs/ctp/src/zi.adoc]: You didn't add anything to ExceptionsSm, so this doesn't seem like it is tested.

**Comment by wangpc-pp** [comment]
Ping!

File: generators/testgen/src/testgen/coverpoints/cr_rs1_imm_edges.py [hunk 10]
```diff
@@ -48,3 +52,58 @@ def make_cr_rs1_imm_edges(instr_name: str, instr_type: str, coverpoint: str, tes
             return_testcase_registers(test_data, params)
 
     return test_chunks
+
+
+def make_cr_rs1_imm_edges_5bit_u_n0_offset(
```
Comment (jordancarlin) [inline generators/testgen/src/testgen/coverpoints/cr_rs1_imm_edges.py]: Move this into a separate file in `coverpoints/special`. Follow the approach taken for `cr_rs1_rs2_edges_offset`.

File: generators/testgen/src/testgen/formatters/types/bi_type.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+    setup = [
+        load_int_reg("rs1", params.rs1, params.rs1val, test_data),
+        load_int_reg("branch taken value", params.temp_reg, params.temp_val, test_data),
+    ]
+
+    test = [
+        (
+            f"{format_zibi_branch(instr_name, params.rs1, params.immval, '1f')} "
+            f"# {instr_name} x{params.rs1}, {params.immval}, 1f; perform operation"
```
Comment (jordancarlin) [inline generators/testgen/src/testgen/formatters/types/bi_type.py]: Why is this using a semicolon instead of a pound to separate the comment?

File: generators/testgen/src/testgen/asm/helpers.py [hunk 12]
```diff
@@ -51,6 +51,27 @@ def to_hex(value: int, bits: int) -> str:
...
+        raise ValueError(f"Unsupported Zibi branch instruction: {instr_name}") from error
+
+    if cimm == -1:
+        cimm_field = 0
+    elif 1 <= cimm <= 31:
+        cimm_field = cimm
+    else:
+        raise ValueError(f"Invalid Zibi comparison immediate: {cimm}")
+
+    return f".insn b 0x63, {funct3}, x{rs1}, x{cimm_field}, {target}"
```
Comment (jordancarlin) [inline generators/testgen/src/testgen/asm/helpers.py]: Drop this function for now and use the actual instruction (even though it doesn't compile). As part of the experimental mode work, I will lay out a convention for macros for unsupported instructions that make the generated assembly much easier to read.

File: generators/testgen/src/testgen/coverpoints/special/cp_offset.py [hunk 13]
```diff
@@ -48,7 +54,21 @@ def make_offset(instr_name: str, instr_type: str, coverpoint: str, test_data: Te
...
+            )
+        else:
+            tc.code.append(
+                f"2: {format_zibi_branch(instr_name, params.rs1, 1, '1b')} "
+                f"# {instr_name} x{params.rs1}, 1, 1b; backward branch"
+            )
+
+        tc.code.extend(
+            [
                 f"addi x{params.temp_reg}, x{params.temp_reg}, -2 # branch not taken, decrement check value",
```
Comment (jordancarlin) [inline generators/testgen/src/testgen/coverpoints/special/cp_offset.py:62]: Once we drop `format_zibi_branch`, I think this can be simplified so it doesn't need to introduce the additional branches.

**Review by jordancarlin — CHANGES_REQUESTED** [review]
Sorry about the delay in reviewing. Here are (hopefully) my final round of comments. I will give it another look once these are addressed and then you should be good to go for freeze. Once the experimental extension support is in place, we can circle back to actually get this merged, but that won't block this extension progressing in the ratification process.

File: docs/ctp/src/zi.adoc (hunk 9 above)
Comment (wangpc-pp) [reply docs/ctp/src/zi.adoc]: Added a Zibi-specific  generator and coverage bins for taken and not-taken / branches targeting PC+6. The normative mapping and CTP description now reference those Zibi-specific coverpoints.

File: config/sail/sail-rv32-max-clang/run_cmd.txt
Comment (wangpc-pp) [reply config/sail/sail-rv32-max-clang/run_cmd.txt:1]: Reverted all four Sail  changes.

File: generators/testgen/src/testgen/formatters/types/bi_type.py (hunk 11 above)
Comment (wangpc-pp) [reply generators/testgen/src/testgen/formatters/types/bi_type.py]: Fixed. Generated Zibi instructions now use normal  comments, with the redundant mnemonic text removed.

File: generators/testgen/src/testgen/coverpoints/special/cp_offset.py (hunk 13 above)
Comment (wangpc-pp) [reply generators/testgen/src/testgen/coverpoints/special/cp_offset.py:62]: Simplified  to use the common instruction-formatting path with BI operand selection; it no longer introduces a Zibi-only branch path.

File: framework/src/act/build_plan.py
Comment (wangpc-pp) [reply framework/src/act/build_plan.py:1]: Reverted the  changes; generic experimental-extension handling is no longer part of this PR.

File: run_tests.py
Comment (wangpc-pp) [reply run_tests.py:1]: Reverted the  changes, so this PR no longer adds the  placeholder.

File: generators/testgen/src/testgen/asm/helpers.py (hunk 12 above)
Comment (wangpc-pp) [reply generators/testgen/src/testgen/asm/helpers.py]: Removed  and now emit the actual / mnemonics throughout the generated tests.

File: docs/DeveloperGuide.md
Comment (wangpc-pp) [reply docs/DeveloperGuide.md:1]: Reverted the  changes.

File: generators/testgen/src/testgen/generate/unpriv.py
Comment (wangpc-pp) [reply generators/testgen/src/testgen/generate/unpriv.py:1]: Reverted the unprivileged generation changes. Zibi now keeps its normal  and  metadata for the future experimental mode.

File: coverpoints/priv/ExceptionsSm_coverage.svh [hunk 14]
```diff
@@ -43,6 +43,18 @@ covergroup ExceptionsSm_cg with function sample(ins_t ins);
...
+        zibi_branches_taken: coverpoint {ins.current.insn[6:0], ins.current.insn[14:12],
+                                         ins.current.rs1_val == ins.current.imm2} {
+            bins beqi_taken = {11'b1100011_010_1};
+            bins bnei_taken = {11'b1100011_011_0};
+        }
+        zibi_branches_nottaken: coverpoint {ins.current.insn[6:0], ins.current.insn[14:12],
+                                            ins.current.rs1_val == ins.current.imm2} {
+            bins beqi_nottaken = {11'b1100011_010_0};
+            bins bnei_nottaken = {11'b1100011_011_1};
+        }
```
Comment (jordancarlin) [inline coverpoints/priv/ExceptionsSm_coverage.svh:56]: For exceptions, we tend to prefer testing in all priv modes. Please also add these coverpoints to ExceptionsS and ExceptionsU.

File: generators/testgen/src/testgen/priv/extensions/ExceptionsZibi.py
Comment (jordancarlin) [inline generators/testgen/src/testgen/priv/extensions/ExceptionsZibi.py:1]: Rather than creating a new ExceptionsZibi test (which won't get coverage for the coverpoint added to ExceptionsSm), let's add the new branches to the existing `generate_instr_adr_misaligned_branch_tests` and `generate_instr_adr_misaligned_branch_nottaken` functions in `ExceptionsCommon.py`. Make sure to guard the new branches with `#ifdef ZIBI_SUPPORTED`. This will automatically add them to ExceptionsSm, ExceptionsS, and ExceptionsU.

**Review by jordancarlin — CHANGES_REQUESTED** [review]
The unpriv tests look great. A few comments on the new priv tests.

**Review by jordancarlin — COMMENTED** [review]
Thanks for all the work on this @wangpc-pp. This looks great and is good to go for the ACT freeze milestone.

I will circle back to this PR once the experimental extensions flag is in place so we can look at merging this.

**Comment by jordancarlin** [comment]
@wangpc-pp I added support for experimental extensions in #2433. Would you be ok if I updated this PR to use that new support to serve as the initial example of an experimental extension?

**Comment by wangpc-pp** [comment]
> Pengcheng Wang (@wangpc-pp) I added support for experimental extensions in #2433. Would you be ok if I updated this PR to use that new support to serve as the initial example of an experimental extension?

Happy to see that! Please go ahead!
